Add combat tracker API

SoSeDiK opened the pull request:
Adds `LivingEntity#getCombatTracker` to get entity's combat tracker that allows getting/setting combat history (list of damage entries during a combat session), recording new combat entries without causing actual damage, getting/resetting current combat state/session and getting a death message based on the present combat history.

`FallLocationType` represents a location from which the entity fell, and is used by vanilla to generate fall death messages.

___

API implementation is basically done, but definitely open to feedback. Javadocs & namings could use some love.

SoSeDiK committed:
Add combat tracker API

kennytv commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/CombatTracker.java:
```diff
@@ -0,0 +1,104 @@
+     * Gets the list of recorded combat entries.
+     * <br>
+     * Mutating the list is safe and won't have
+     * an effect on this entity's combat history.
```
It should just be immutable/say it's a copy

SoSeDiK replied:
Updated the wording to state it's a copy.
Haven't made this immutable because it's returning a copy internally each time, so modifications are safe. Otherwise, users would need to make a copy of a copy.

kennytv commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/CombatTracker.java:
```diff
@@ -0,0 +1,104 @@
+     * @see #resetCombatState()
+     * @see #addCombatEntry(CombatEntry)
+     */
+    void setCombatEntries(List<CombatEntry> combatEntries);
```
What's the use of this?

SoSeDiK replied:
"Sneakily" setting prior combat history without overriding/resetting active combat state.
It's a neat shortcut for providing fake pre-made combat history scenarios and producing specific death messages.

kennytv commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/FallLocation.java:
```diff
@@ -0,0 +1,65 @@
+ */
+@NullMarked
+@ApiStatus.Experimental
+public sealed interface FallLocation extends Translatable permits FallLocationImpl {
```
FallLocationType maybe? Rn it sounds like this is actually holding a location, the vanilla name doesn't really work in our API

SoSeDiK replied:
Yes, the Type makes more sense

SoSeDiK committed:
Apply requested changes

- Rename FallLocation to FallLocationType
- Change CombatTracker#getCombatEntries javadoc

SoSeDiK committed:
Fixup missed fall location types

lynxplay commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/CombatEntry.java:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Represents a combat entry
+ *
+ * @since 1.21.4
```
We have for now decided against since annotation in the project.
Adding *some* is just gonna make our lifes even more miserable when/if we solve them semi automatically.

lynxplay commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/CombatTracker.java:
```diff
@@ -0,0 +1,106 @@
+     *
+     * @return the most significant fall damage entry
+     */
+    @Nullable CombatEntry getMostSignificantFall();
```
computeMostSignificantFall

lynxplay commented on paper-server/patches/sources/net/minecraft/world/damagesource/CombatTracker.java.patch:
```diff
@@ -0,0 +1,48 @@
++    }
++
++    public void resetCombatState() {
++        // Paper end - Combat tracker API
```
The intend ruined your diff here.

lynxplay commented on paper-server/src/main/java/io/papermc/paper/PaperServerInternalAPIBridge.java:
```diff
@@ -12,4 +21,17 @@ public class PaperServerInternalAPIBridge implements InternalAPIBridge {
+        net.minecraft.world.entity.LivingEntity mob = ((CraftLivingEntity) entity).getHandle();
+        net.minecraft.world.damagesource.FallLocation fallLocation = net.minecraft.world.damagesource.FallLocation.getCurrentFallLocation(mob);
+        float fallDistance = mob.fallDistance;
+        return new PaperCombatEntryWrapper(new net.minecraft.world.damagesource.CombatEntry(((CraftDamageSource) damageSource).getHandle(), damage, fallLocation, fallDistance));
```
can you just call the `createCombatEntry` overload here.
Makes life easier if that ctor ever changes.

lynxplay commented on paper-server/src/main/java/io/papermc/paper/world/damagesource/PaperCombatTrackerWrapper.java:
```diff
@@ -0,0 +1,114 @@
+
+    @Override
+    public List<CombatEntry> getCombatEntries() {
+        List<CombatEntry> combatEntries = new ArrayList<>();
```
Pre-allocate with size.

SoSeDiK committed:
Address the requested changes

Lulu13022002 commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/FallLocationType.java:
```diff
@@ -0,0 +1,63 @@
+     */
+    FallLocationType SCAFFOLDING = new FallLocationTypeImpl("scaffolding");
+    /**
+     * The entity is within some other climable block
```
Small typo on climable

Lulu13022002 commented on paper-api/src/main/java/io/papermc/paper/world/damagesource/CombatEntry.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return the fall location
+     */
+    @Nullable FallLocationType getFallLocation();
```
Did you forget to rename this one too? same in the javadocs here and below

Lulu13022002 commented on build-data/paper.at:
```diff
@@ -713,6 +714,10 @@ public-f net.minecraft.server.dedicated.DedicatedServer settings
+public-f net.minecraft.world.damagesource.CombatTracker entries
+public-f net.minecraft.world.damagesource.CombatTracker inCombat
+public-f net.minecraft.world.damagesource.CombatTracker mob
+public-f net.minecraft.world.damagesource.CombatTracker takingDamage
```
There's no need to remove the final modifier for those?

SoSeDiK committed:
Address the requested changes

Warriorrrr committed:
Replace line break with paragraph

lynxplay commented on paper-server/patches/sources/net/minecraft/world/damagesource/CombatTracker.java.patch:
```diff
@@ -0,0 +1,37 @@
+             }
+ 
+             this.entries.clear();
+-        }
```
Add a second opening bracket to the method start instead.

lynxplay committed:
Fix merge conflict and styling

lynxplay committed:
Optionull

lynxplay commented on paper-server/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java:
```diff
@@ -1167,4 +1174,15 @@ public void setBodyYaw(final float bodyYaw) {
+    }
+
+    @Override
+    public FallLocationType calculateFallLocationType() {
```
I am a bit confused on these.
The entry yields this nullable, which mirrors internals, while this falls back to GENERIC.
What is the line of thought here?

SoSeDiK replied:
I did what vanilla does when calculating the fall message:
```java
FallLocation fallLocation = Objects.requireNonNullElse(combatEntry.fallLocation(), FallLocation.GENERIC);
```
The reasoning is, since this calculates the fall location type, then if there's no special fall type, yield `GENERIC` (generic is only used internally as a fallback when the type is `null` iirc, so it was `null` vs `GENERIC`, and I liked the latter more)
But if I've missed something, nullable is also fine

lynxplay replied:
Yea I know it is used internally for the death message computation, but that seems to be about it in terms of usage for that type?
The disconnect between "LivingEntity#calculateFallLocationType" and "create new combat entry from living entity -> ah it returns null here now eventho it just yielded GENERIC" is a bit unfortunate imo.
Tho I guess that is more on mojang than on this API given they do the "fallback" so late in their logic.
It is fine like this then, tho maybe a bit of javadocs if you have some time.

SoSeDiK replied:
> it returns null here now eventho it just yielded GENERIC"

It's quite ambiguous (javadoc suggestions?), but is expected from my point of view
Having `GENERIC` in the entry represents a fall (and this method is meant for that), while `null` means that the entry is not a fall damage one

lynxplay replied:
Yea but `calculateFallLocationType` yielding GENERIC then, when the entity isn't falling rn is kind of a counterpoint.
Having a non null falltype in the entry (at least according to internals) does not define if that was a fall.
That is tracked (correctly so) via the DamageSource.
You can be in the air and killed by an arrow, that kill won't be a registered as a fall related death even though the last entry does hold a falllocation type.

Javadocs on CombatEntry#combatEntry also currently link to the calc method, which then would not be correct right?

SoSeDiK replied:
Good points, definitely need to at least clarify this in javadocs

lynxplay committed:
Move fall location computation to CombatTracker

lynxplay committed:
Fix rebase

lynxplay committed:
Fix javadocs
